### PR TITLE
Message presneter

### PR DESCRIPTION
**Pull request by falkzach**
A WIP for a presentation layer, seperating component rendering from the client. A presenter class is used to render content in different formats.

@benjic what are your thoughts of this so far?

### CODE REVIEW

File: lib/src/presenters/message_presenter_factory.dart [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import 'package:client/src/messages/markdown.dart';
+import 'package:client/src/messages/link.dart';
+import 'package:client/src/messages/file.dart';
+
+abstract class PresenterFactory {
+  MessagePresenter getPresenter(Message);
+}
+
+class MessagePresenterFactory {
+  MessagePresenter getPresenter(Message m) {
```
Comment (benjic) [inline lib/src/presenters/message_presenter_factory.dart]: Can this be static?

File: lib/src/messages/link.dart [hunk 2]
```diff
@@ -13,4 +18,7 @@ class Link extends Message {
   Map marshal() {
     return new Map();
   }
+
+  String getTitle() => title;
+  Uri getRef() => ref;
```
Comment (benjic) [inline lib/src/messages/link.dart]: Dart has getter support for class properties:

```dart
class Link extends Message {
    String _title;
    Uri _ref;

    String get title => _title;
    Uri get ref => _ref;
}
```

File: lib/src/presenters/message_preseneter.dart [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  @override
+  String presentPlaintext(Message m) {
+    // TODO: implement presentPlaintext
+  }
+}
+
+class LinkPresenter extends MessagePresenter
+{
+  @override
+  String presentHTML(Link l) => '<a href="' + l.getRef().toString() + '">' + l.getTitle() + '</a>';
```
Comment (benjic) [inline lib/src/presenters/message_preseneter.dart]: Dart has string interpolation:

```dart
String present(Link l) => '<a href="${l.ref}">${l.title}</a>';
```

This uses the getters defined in my example above.

File: test/presenters/message_presenter_test.dart [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+          equals('[Example Domain](http://example.com)'));
+    });
+
+    test('should generage plain text link', () {
+      expect(presenter.presentPlaintext(link),
+          equals('Example Domain\t-\thttp://example.com'));
+    });
+
+  });
+}
```
Comment (benjic) [inline test/presenters/message_presenter_test.dart]: Great tests!

File: lib/src/presenters/message_presenter_factory.dart (hunk 1 above)
Comment (falkzach) [inline lib/src/presenters/message_presenter_factory.dart]: I believe so yes

**Comment by benjic** [comment]
Would it be worthwhile to define the abstract `Presenter` and presenter factories within the core library and allow consumers to pass a concrete instance as a form of dependency injection? When messages are unmarshalled the passed factory would create a presenter instance which it can call and pass `this`. This would allow the actual web application define how to render a message without the library knowing directly how.

This would simplify the presenter interface and would prevent having to know all presentation formats now.

**Comment by benjic** [comment]
This attempt has been invalidated.

**Pull request closed without merging**